src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/AddVenueController.cs

```diff
@@ -0,0 +1,193 @@
+
+            return await _mediator.SendAndMapResponse(
+                command,
+                result => result.Match<IActionResult>(
+                    errors => this.ViewFromErrors(errors),
+                    success => RedirectToAction(nameof(CheckAndPublish))
+                        .WithProviderContext(_providerContextProvider.GetProviderContext())
+                        .WithJourneyInstanceUniqueKey(_journeyInstance)));
+        }
+
+        [HttpGet("check-publish")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> CheckAndPublish() => await _mediator.SendAndMapResponse(
+            new CheckAndPublish.Query(),
+            vm => View(vm));
+
+        [HttpPost("publish")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> Publish(CheckAndPublish.Command command) =>
+            await _mediator.SendAndMapResponse(
+                command,
+                success =>
+                {
+                    // LEGACY: Some legacy journeys display a notification when a venue has been added.
+                    // Stash the venue's address in TempData so it can be read by those legacy views.
+                    TempData.Add(
+                        TempDataKeys.AddedVenueDescription,
+                        string.Join(
+                            ", ",
+                            new[]
+                            {
+                                _journeyInstance.State.Name,
+                                _journeyInstance.State.AddressLine1,
+                                _journeyInstance.State.AddressLine2,
+                                _journeyInstance.State.Town,
+                                _journeyInstance.State.County,
+                                _journeyInstance.State.Postcode
+                            }
+                            .Where(l => !string.IsNullOrWhiteSpace(l))));
+
+                    if (_journeyInstance.Properties.TryGetValue(ReturnUrlJourneyInstancePropertyKey, out var returnUrlObj) &&
+                        returnUrlObj is string returnUrl &&
+                        Url.IsLocalUrl(returnUrl))
+                    {
+                        var venueId = _journeyInstance.State.VenueId.Value;
+                        return Redirect(new Url(returnUrl).SetQueryParam("venueId", venueId));
+                    }
+
+                    return (IActionResult)RedirectToAction(nameof(Published))
+                        .WithProviderContext(_providerContextProvider.GetProviderContext())
+                        .WithJourneyInstanceUniqueKey(_journeyInstance);
+                });
+
+        [HttpGet("success")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> Published()
+        {
+            var query = new Published.Query();
+            return await _mediator.SendAndMapResponse(query, vm => View(vm));
```

Comment:
No need to await this! (or many of the other Actions in this class, but this is the obvious one that can be made more succinct)

Reply:
Can't just `return` here - the `Task<IViewResult>` isn't assignable to `Task<IActionResult>`

Reply:
```
public Task<IActionResult> Published() =>
    _mediator.SendAndMapResponse<Published.ViewModel, IActionResult>(new Published.Query(), vm => View(vm));
```

Reply:
That's less readable to my eyes.